sw/host/opentitanlib/src/bootstrap/primitive.rs

```diff
@@ -0,0 +1,160 @@
+    const DATA_LEN: usize = 2048 - std::mem::size_of::<FrameHeader>();
+    const HASH_LEN: usize = 32;
+
+    /// Computes the hash in the header.
+    fn header_hash(&self) -> [u8; Frame::HASH_LEN] {
+        let frame = self.as_bytes();
+        let sha = Sha256::hash(&frame[Frame::HASH_LEN..]);
+        sha.bytes()
+    }
+
+    /// Computes the hash over the entire frame.
+    fn frame_hash(&self) -> [u8; Frame::HASH_LEN] {
+        let sha = Sha256::hash(self.as_bytes());
+        sha.bytes()
+    }
+
+    /// Creates a sequence of frames based on a `payload` binary.
+    fn from_payload(payload: &[u8]) -> Vec<Frame> {
+        let mut frames = Vec::new();
+        let last_frame = (payload.len() + Frame::DATA_LEN - 1) / Frame::DATA_LEN - 1;
+        for (i, chunk) in payload.chunks(Frame::DATA_LEN).enumerate() {
+            let mut frame = Frame::default();
+            frame.header.frame_num = i as u32;
+            if i == last_frame {
+                frame.header.frame_num |= Frame::EOF;
+            }
+            frame.header.flash_offset = (i * Frame::DATA_LEN) as u32;
+            frame.data[..chunk.len()].copy_from_slice(chunk);
+            frame.header.hash = frame.header_hash();
+            frames.push(frame)
+        }
+        frames
+    }
+}
+
+/// Implements the primitive bootstrap protocol.
+pub struct Primitive {
+    pub inter_frame_delay: Duration,
+    pub flash_erase_delay: Duration,
+}
+
+impl Primitive {
+    const INTER_FRAME_DELAY: Duration = Duration::from_millis(20);
+    const FLASH_ERASE_DELAY: Duration = Duration::from_millis(200);
+
+    /// Creates a new `Primitive` protocol updater from `options`.
+    pub fn new(options: &BootstrapOptions) -> Self {
+        Primitive {
+            inter_frame_delay: options.inter_frame_delay.unwrap_or(Self::INTER_FRAME_DELAY),
+            flash_erase_delay: options.flash_erase_delay.unwrap_or(Self::FLASH_ERASE_DELAY),
+        }
+    }
+}
+
+impl UpdateProtocol for Primitive {
+    /// Performs the update protocol using the `transport` with the firmware `payload`.
+    fn update(&self, spi: &dyn Target, payload: &[u8]) -> Result<()> {
+        let frames = Frame::from_payload(payload);
+
+        let mut i = 0;
+        while i < frames.len() {
+            let frame = &frames[i];
+            log::info!(
+                "Writing frame {} (offset {:x?})",
+                i,
+                frame.header.flash_offset
+            );
+
+            // Write the frame and read back the hash of the previous frame.
+            let mut prev_hash = [0u8; std::mem::size_of::<Frame>()];
+            spi.run_transaction(&mut [Transfer::Both(frame.as_bytes(), &mut prev_hash)])?;
+
+            if i == 0 {
+                // If its the first frame, there is no hash to check.
+                // We need to give the target some time to erase the flash.
+                std::thread::sleep(self.flash_erase_delay);
+                i += 1;
+            } else {
```

Comment:
I'd get rid of the else and put a continue at the end of the first block; same for the `if` below. (or maybe it makes sense to do it as `if != want_hash { error!(); continue }`...

Reply:
Done.